torchao/quantization/qat/fake_quantize_config.py

```diff
@@ -0,0 +1,262 @@
+    """
+    Config for how to fake quantize weights or activations.
+
+    Args:
+        dtype: dtype to simulate during fake quantization, e.g. torch.int8.
+            For PyTorch versions older than 2.6, you may use `TorchAODType` to represent
+            torch.int1 to torch.int7 instead, e.g. TorchAODType.INT4.
+        granularity: granularity of scales and zero points, e.g. PerGroup(32).
+            We also support the following strings:
+               1) 'per_token': equivalent to PerToken()
+               2) 'per_channel': equivalent to PerAxis(0)
+               3) 'per_group': equivalent to PerGroup(group_size), must be combined
+                   with separate `group_size` kwarg, Alternatively, just set the
+                   `group_size` kwarg and leave this field empty.
+        mapping_type: whether to use symmetric (default) or asymmetric quantization
+            Alternatively, set `is_symmetric` (bool) and leave this field empty.
+        scale_precision: scale dtype (default torch.fp32)
+        zero_point_precision: zero point dtype (default torch.int32)
+        zero_point_domain: whether zero point is in integer (default) or float domain
+        is_dynamic: whether to use dynamic (default) or static scale and zero points
+        range_learning (prototype): whether to learn scale and zero points during training
+            (default false), not compatible with `is_dynamic`.
+
+    Keyword args:
+        group_size: size of each group in per group fake quantization,
+            can be set instead of `granularity`
+        is_symmetric: whether to use symmetric or asymmetric quantization,
+            can be set instead of `mapping_type`
+
+    Example usage::
+
+        # Per token asymmetric quantization
+        IntxFakeQuantizeConfig(torch.int8, "per_token", is_symmetric=False)
+        IntxFakeQuantizeConfig(torch.int8, PerToken(), MappingType.ASYMMETRIC)
+
+        # Per channel symmetric quantization
+        IntxFakeQuantizeConfig(torch.int4, "per_channel")
+        IntxFakeQuantizeConfig(torch.int4, "per_channel", is_symmetric=True)
+        IntxFakeQuantizeConfig(torch.int4, PerAxis(0), MappingType.SYMMETRIC)
+
+        # Per group symmetric quantization
+        IntxFakeQuantizeConfig(torch.int4, group_size=32)
+        IntxFakeQuantizeConfig(torch.int4, group_size=32, is_symmetric=True)
+        IntxFakeQuantizeConfig(torch.int4, "per_group", group_size=32, is_symmetric=True)
+        IntxFakeQuantizeConfig(torch.int4, PerGroup(32), MappingType.SYMMETRIC)
+    """
+
+    dtype: Union[torch.dtype, TorchAODType]
+    granularity: Granularity
+    mapping_type: MappingType
+    scale_precision: torch.dtype
+    zero_point_precision: torch.dtype
+    zero_point_domain: ZeroPointDomain
+    is_dynamic: bool = True
+    range_learning: bool = False
+    eps: Optional[float] = None
+
+    def __init__(
+        self,
+        dtype: Union[torch.dtype, TorchAODType],
```

Comment:
I wonder if you can type this as `Literal[...]` so that it only allows for int inputs

Reply:
just define a `Allowed Types` above and use it

Reply:
Hmm I just tried it but didn't really like it. I think I prefer a simpler signature like just `torch.dtype` (we can drop `TorchAODType` soon, only needed for pytorch 2.5 and before) and do the validation in init. Also this is just moving existing code

Reply:
Hmm I just tried it but didn't really like it. I think I prefer a simpler signature like just `torch.dtype` (we can drop `TorchAODType` soon, only needed for pytorch 2.5 and before) and do the validation in init